### ISSUE TITLE
counterexample for tinterp_2

### ISSUE THREAD

**Issue by farzadhtg** [youqad/Coherence-of-Heyting-arithmetic#6]
I think tinterp_2 is not true. Consider the following example:

length v1= 5
length v2 = 3
t = Tvar 6
t' = Tvar 2

So, we have: 

tsubst 5 Tvar 2 Tvar 6 = Tvar (pred 6) = Tvar 5

and then: 

tinterp (v1 ++ v2) (Tvar 5) = nth 5 (v1++v2) 0  = v1[5]

on the other hand: 

tinterp v2 Tvar 2 = nth 2 v2 0

and then:

tinterp (v1 ++ (nth 2 v2 0) :: v2) Tvar 6 = nth 6 (v1 ++ (nth 2 v2 0) :: v2) 0 = nth 2 v2 = v2[2]

which v1[5] \neq v2[2].

Have I misunderstood something?

**Comment by AliceRixte** [comment]
OK I'm not really sure but I thins you count starting from 1 and I think this is the mistake : nth 5 = v2[0]

EDIT : the more I read it the less I'm confident about what I wrote.

**Comment by AliceRixte** [comment]
I think I get it. I think the problem is when you use a varible which is not in v1. To fix it we must add 

> cterm t (length v1)

**Comment by farzadhtg** [comment]
Yes, you're right. But I don't see how are you fixing it. Do you mean we should change the statement of the lemma?

**Comment by AliceRixte** [comment]
Yes the lemma would be
```

Lemma tinterp_2 : forall t' t v1 v2, length v2 = 1 ->
  tinterp (v1 ++ v2) (tsubst (length v1) t' t) =
tinterp (v1 ++ (tinterp v2 t') :: v2) t.
```

**Comment by farzadhtg** [comment]
this is same as existence code. Am I right? what is difference?

**Comment by AliceRixte** [comment]
what is existence code?
Edit : my bad I get it ...
 I added
`length v2 = 1 ->`

**Comment by youqad** [comment]
@farzadhtg I don't think you yielded a counterexample. In case

```coq
length v1= 5
length v2 = 3
t = Tvar 6
t' = Tvar 2
```
then indeed

```coq
tsubst 5 (Tvar 2) (Tvar 6) = Tvar (pred 6) = Tvar 5
```

but, as the indices of `v1` range from 0 to 4  = (length v1)-1:

```coq
tinterp (v1 ++ v2) (Tvar 5) = nth 5 (v1++v2) 0 = v2[0]
```

And on the flipside of the coin:

```coq
tinterp v2 (Tvar 2) = nth 2 v2 0 = v2[2]
```

indeed, as a result of which:

```coq
tinterp (v1 ++ v2[2] :: v2) (Tvar 6) = (v1 ++ v2[2] :: v2)[6] = (v2[2] :: v2)[1] = v2[0]
```
